File: shared/python/async_bridge_client.py
"""
Stitch - asyncio-based Python bridge client.

Provides AsyncBridgeClientBase for use with async Python frameworks
(FastAPI, aiohttp, Starlette, etc.).

Usage:
    class MyBridge(AsyncBridgeClientBase):
        async def add(self, a: float, b: float) -> dict:
            return await self._call("add", {"a": a, "b": b})

    async with MyBridge(["ruby", "sidecar.rb"]) as bridge:
        result = await bridge.add(1, 2)
"""

from __future__ import annotations

import asyncio
import json
import os
import signal
import sys
import uuid
from typing import Any, AsyncGenerator


class AsyncBridgeError(Exception):
    """Raised when the remote sidecar returns a JSON-RPC error object."""

    def __init__(self, code: int | None, message: str) -> None:
        super().__init__(f"[{code}] {message}")
        self.code = code
        self.message = message


class AsyncBridgeClientBase:
    """
    Abstract base for asyncio-based Python bridge clients.

    Subclasses must call super().__init__(cmd) and may add typed async public
    methods that delegate to self._call(method, params).

    Usage (async context manager - preferred):

        class MyBridge(AsyncBridgeClientBase):
            async def add(self, a, b):
                return await self._call("add", {"a": a, "b": b})

        async with MyBridge(["ruby", "sidecar.rb"]) as bridge:
            result = await bridge.add(1, 2)

    Usage (manual):

        bridge = MyBridge(["./sidecar"])
        await bridge.start()
        result = await bridge._call("echo", {"msg": "hi"})
        await bridge.stop()
    """

    READY_TIMEOUT: float = 10.0
    CALL_TIMEOUT: float = 30.0

    def __init__(
        self,
        cmd: list[str],
        ready_timeout: float | None = None,
        call_timeout: float | None = None,
        env: dict[str, str] | None = None,
    ) -> None:
        self._cmd = cmd
        self._ready_timeout = ready_timeout if ready_timeout is not None else self.READY_TIMEOUT
        self._call_timeout = call_timeout if call_timeout is not None else self.CALL_TIMEOUT
        self._env = env

        self._proc: asyncio.subprocess.Process | None = None
        self._reader_task: asyncio.Task | None = None

        # id -> asyncio.Future[dict]
        self._pending: dict[str, asyncio.Future] = {}

        # id -> asyncio.Queue for streaming
        self._stream_pending: dict[str, asyncio.Queue] = {}

        self._ready_event: asyncio.Event | None = None
        self._closed = False

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    async def start(self) -> None:
        """Spawn the sidecar process and wait for the {"ready":true} handshake."""
        if self._proc is not None:
            return

        env = {**os.environ, **(self._env or {})}

        self._proc = await asyncio.create_subprocess_exec(
            *self._cmd,
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
            env=env,
        )

        self._closed = False
        self._ready_event = asyncio.Event()

        self._reader_task = asyncio.create_task(
            self._reader_loop(), name="async-bridge-reader"
        )

        # Drain stderr in the background.
        asyncio.create_task(self._stderr_drain(), name="async-bridge-stderr")

        try:
            await asyncio.wait_for(
                self._ready_event.wait(), timeout=self._ready_timeout
            )
        except asyncio.TimeoutError:
            await self.stop(force=True)
            raise TimeoutError(
                f'Sidecar did not emit {{"ready":true}} within {self._ready_timeout}s. '
                f"Command: {self._cmd}"
            )

    async def stop(self, force: bool = False) -> None:
        """Gracefully terminate the child process."""
        if self._closed:
            return
        self._closed = True

        # Wake all pending callers with an error.
        for fut in self._pending.values():
            if not fut.done():
                fut.set_exception(
                    AsyncBridgeError(-32000, "bridge closed")
                )
        self._pending.clear()

        # Wake all stream waiters.
        for sq in self._stream_pending.values():
            await sq.put(("error", {"code": -32000, "message": "bridge closed"}))
        self._stream_pending.clear()

        if self._reader_task is not None:
            self._reader_task.cancel()
            try:
                await self._reader_task
            except (asyncio.CancelledError, Exception):
                pass
            self._reader_task = None

        if self._proc is None:
            return

        try:
            if self._proc.returncode is None:
                if not force:
                    self._proc.terminate()
                    try:
                        await asyncio.wait_for(self._proc.wait(), timeout=2.0)
                    except asyncio.TimeoutError:
                        pass
                if self._proc.returncode is None:
                    self._proc.kill()
                    await self._proc.wait()
        except (ProcessLookupError, OSError):
            pass

        self._proc = None

    # ------------------------------------------------------------------
    # Context manager
    # ------------------------------------------------------------------

    async def __aenter__(self) -> "AsyncBridgeClientBase":
        await self.start()
        return self

    async def __aexit__(self, *_: Any) -> None:
        await self.stop()

    # ------------------------------------------------------------------
    # Built-in convenience methods
    # ------------------------------------------------------------------

    async def ping(self) -> dict:
        """Send a built-in __ping__ call. Returns {"pong": True, "pid": <int>}."""
        return await self._call("__ping__", {}, timeout=5.0)

    # ------------------------------------------------------------------
    # Protected RPC primitives
    # ------------------------------------------------------------------

    async def _call(
        self,
        method: str,
        params: dict[str, Any] | None = None,
        timeout: float | None = None,
        traceparent: str | None = None,
    ) -> Any:
        """
        Send a JSON-RPC request and await the response.

        Returns the ``result`` field on success.
        Raises :class:`AsyncBridgeError` if the sidecar returns an error object.
        Raises :class:`TimeoutError` if no response arrives within the timeout.
        """
        if self._closed or self._proc is None:
            raise RuntimeError("Bridge is not running - call start() first")

        call_id = str(uuid.uuid4())
        loop = asyncio.get_running_loop()
        fut: asyncio.Future = loop.create_future()

        self._pending[call_id] = fut

        request: dict[str, Any] = {
            "id": call_id,
            "method": method,
            "params": params or {},
        }
        if traceparent is not None:
            request["traceparent"] = traceparent

        await self._send_raw(json.dumps(request, separators=(",", ":")) + "\n")

        deadline = timeout if timeout is not None else self._call_timeout
        try:
            return await asyncio.wait_for(asyncio.shield(fut), timeout=deadline)
        except asyncio.TimeoutError:
            self._pending.pop(call_id, None)
            raise TimeoutError(f"No response for method={method!r} within {deadline}s")

    async def stream(
        self,
        method: str,
        params: dict[str, Any] | None = None,
        timeout: float | None = None,
        traceparent: str | None = None,
    ) -> AsyncGenerator[Any, None]:
        """
        Send a streaming JSON-RPC call. Yields each chunk as it arrives.
        The sidecar handler must return a StreamResponse wrapping a generator.
        Raises AsyncBridgeError on remote error, TimeoutError on chunk timeout.
        """
        if self._closed or self._proc is None:
            raise RuntimeError("Bridge is not running")

        call_id = str(uuid.uuid4())
        q: asyncio.Queue = asyncio.Queue()
        self._stream_pending[call_id] = q

        request: dict[str, Any] = {
            "id": call_id,
            "method": method,
            "params": params or {},
            "stream": True,
        }
        if traceparent is not None:
            request["traceparent"] = traceparent

        await self._send_raw(json.dumps(request, separators=(",", ":")) + "\n")

        deadline = timeout if timeout is not None else self._call_timeout

        try:
            while True:
                try:
                    kind, value = await asyncio.wait_for(q.get(), timeout=deadline)
                except asyncio.TimeoutError:
                    self._stream_pending.pop(call_id, None)
                    raise TimeoutError(
                        f"Stream timeout for method={method!r} after {deadline}s"
                    )
                if kind == "done":
                    return
                elif kind == "error":
                    raise AsyncBridgeError(
                        value.get("code"), value.get("message", "stream error")
                    )
                else:  # chunk
                    yield value
        finally:
            self._stream_pending.pop(call_id, None)

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    async def _send_raw(self, line: str) -> None:
        if self._proc is None or self._proc.stdin is None:
            raise RuntimeError("Child stdin is not available")
        encoded = line.encode("utf-8")
        self._proc.stdin.write(encoded)
        try:
            await self._proc.stdin.drain()
        except (BrokenPipeError, ConnectionResetError, OSError) as exc:
            raise RuntimeError("Child stdin pipe is broken") from exc

    async def _reader_loop(self) -> None:
        assert self._proc is not None
        stdout = self._proc.stdout
        if stdout is None:
            return

        try:
            while True:
                try:
                    raw_line = await stdout.readline()
                except asyncio.CancelledError:
                    raise
                except Exception:
                    break

                if not raw_line:
                    break  # EOF

                line = raw_line.decode("utf-8", errors="replace").strip()
                if not line:
                    continue

                try:
                    msg: dict[str, Any] = json.loads(line)
                except json.JSONDecodeError:
                    continue

                # Ready handshake
                if msg.get("ready") is True:
                    if self._ready_event is not None:
                        self._ready_event.set()
                    continue

                # Streaming chunk frame
                if "chunk" in msg and "result" not in msg and "error" not in msg:
                    call_id = msg.get("id")
                    sq = self._stream_pending.get(call_id)
                    if sq is not None:
                        await sq.put(("chunk", msg["chunk"]))
                    continue

                # RPC response dispatch
                call_id = msg.get("id")
                if call_id is None:
                    continue

                # Terminal frame for a stream call
                sq = self._stream_pending.get(call_id)
                if sq is not None:
                    self._stream_pending.pop(call_id, None)
                    if "error" in msg:
                        await sq.put(("error", msg["error"]))
                    else:
                        await sq.put(("done", None))
                    continue

                # Regular RPC future
                fut = self._pending.pop(call_id, None)
                if fut is None or fut.done():
                    continue

                if "error" in msg:
                    err = msg["error"]
                    fut.set_exception(
                        AsyncBridgeError(err.get("code"), err.get("message", "unknown error"))
                    )
                else:
                    fut.set_result(msg.get("result"))

        except asyncio.CancelledError:
            pass
        finally:
            # EOF or cancelled - wake any remaining waiters
            for fut in self._pending.values():
                if not fut.done():
                    fut.set_exception(
                        AsyncBridgeError(-32000, "child process exited")
                    )
            self._pending.clear()

            for sq in self._stream_pending.values():
                await sq.put(("error", {"code": -32000, "message": "child process exited"}))
            self._stream_pending.clear()

            # Prevent start() from hanging if child dies before ready.
            if self._ready_event is not None and not self._ready_event.is_set():
                self._ready_event.set()

    async def _stderr_drain(self) -> None:
        if self._proc is None or self._proc.stderr is None:
            return
        try:
            while True:
                chunk = await self._proc.stderr.read(4096)
                if not chunk:
                    break
        except (asyncio.CancelledError, Exception):
            pass

File: shared/rust/async_bridge_client.rs
//! Stitch - async Rust bridge client (Tokio).
//!
//! Requires the following in `Cargo.toml`:
//! ```toml
//! [dependencies]
//! tokio = { version = "1", features = ["full"] }
//! tokio-util = { version = "0.7", features = ["codec"] }
//! serde = { version = "1", features = ["derive"] }
//! serde_json = "1"
//! uuid = { version = "1", features = ["v4"] }
//! ```
//!
//! # Example
//! ```rust,no_run
//! #[tokio::main]
//! async fn main() {
//!     let mut bridge = AsyncBridge::spawn("python", &["sidecar.py"]).await.unwrap();
//!     let result = bridge.call("add", serde_json::json!({"a": 1, "b": 2})).await.unwrap();
//!     println!("{result}");
//!
//!     let mut stream = bridge.stream("generate", serde_json::json!({"prompt": "hi"})).await.unwrap();
//!     while let Some(chunk) = stream.next().await {
//!         println!("{}", chunk.unwrap());
//!     }
//!
//!     bridge.close().await;
//! }
//! ```

use std::collections::HashMap;
use std::sync::Arc;

use serde_json::Value;
use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
use tokio::process::{Child, ChildStdin, Command};
use tokio::sync::{
    mpsc::{self, UnboundedSender},
    oneshot, Mutex,
};

// ─────────────────────────────────────────────────────────────────────────────
// Error types
// ─────────────────────────────────────────────────────────────────────────────

/// A JSON-RPC error returned by the sidecar process.
#[derive(Debug, Clone)]
pub struct AsyncRpcError {
    pub code: Option<i64>,
    pub message: String,
    pub traceback: Option<String>,
    pub backtrace: Option<String>,
}

impl std::fmt::Display for AsyncRpcError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.message)
    }
}

impl std::error::Error for AsyncRpcError {}

impl From<AsyncRpcError> for String {
    fn from(e: AsyncRpcError) -> Self {
        e.message
    }
}

fn parse_rpc_error(v: &Value) -> AsyncRpcError {
    AsyncRpcError {
        code: v.get("code").and_then(Value::as_i64),
        message: v
            .get("message")
            .and_then(Value::as_str)
            .unwrap_or("unknown error")
            .to_string(),
        traceback: v
            .get("traceback")
            .and_then(Value::as_str)
            .map(String::from),
        backtrace: v
            .get("backtrace")
            .and_then(Value::as_str)
            .map(String::from),
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// Internal dispatch tables
// ─────────────────────────────────────────────────────────────────────────────

/// Oneshot sender for regular (non-streaming) calls.
type CallSender = oneshot::Sender<Result<Value, AsyncRpcError>>;

/// Unbounded sender for streaming calls.  The reader task drops the sender
/// (by removing it from the map) on the terminal / error frame, which closes
/// the channel and signals EOF to the consumer.
type StreamSender = UnboundedSender<Result<Value, AsyncRpcError>>;

#[derive(Default)]
struct Dispatch {
    calls: HashMap<String, CallSender>,
    streams: HashMap<String, StreamSender>,
}

type SharedDispatch = Arc<Mutex<Dispatch>>;

// ─────────────────────────────────────────────────────────────────────────────
// AsyncStream - manual async iterator over stream chunks
// ─────────────────────────────────────────────────────────────────────────────

/// An async stream of chunks returned by [`AsyncBridge::stream`].
///
/// Poll with `.next().await` until `None` is returned (end of stream).
pub struct AsyncStream {
    rx: mpsc::UnboundedReceiver<Result<Value, AsyncRpcError>>,
}

impl AsyncStream {
    /// Returns the next chunk, or `None` when the stream is exhausted.
    pub async fn next(&mut self) -> Option<Result<Value, AsyncRpcError>> {
        self.rx.recv().await
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// AsyncBridge
// ─────────────────────────────────────────────────────────────────────────────

/// An async Tokio-based bridge client.
///
/// Spawns a sidecar child process and communicates with it over
/// newline-delimited JSON on stdin / stdout.
pub struct AsyncBridge {
    stdin: ChildStdin,
    child: Child,
    dispatch: SharedDispatch,
    /// Background reader task handle - kept alive for the lifetime of the bridge.
    _reader: tokio::task::JoinHandle<()>,
}

impl AsyncBridge {
    // ── Construction ────────────────────────────────────────────────────────

    /// Spawn `program` with `args`, wait for the `{"ready":true}` handshake,
    /// and return a connected `AsyncBridge`.
    pub async fn spawn(program: &str, args: &[&str]) -> Result<Self, String> {
        let mut child = Command::new(program)
            .args(args)
            .stdin(std::process::Stdio::piped())
            .stdout(std::process::Stdio::piped())
            .stderr(std::process::Stdio::inherit())
            .spawn()
            .map_err(|e| format!("failed to spawn '{program}': {e}"))?;

        let stdin = child.stdin.take().ok_or("no stdin")?;
        let stdout = child.stdout.take().ok_or("no stdout")?;

        let dispatch: SharedDispatch = Arc::new(Mutex::new(Dispatch::default()));

        // ready channel - signalled by the reader task when it sees {"ready":true}
        let (ready_tx, ready_rx) = oneshot::channel::<()>();

        let reader_dispatch = Arc::clone(&dispatch);
        let _reader = tokio::spawn(Self::reader_loop(
            BufReader::new(stdout),
            reader_dispatch,
            ready_tx,
        ));

        // Wait for the ready handshake (5-second timeout)
        tokio::time::timeout(std::time::Duration::from_secs(5), ready_rx)
            .await
            .map_err(|_| "timed out waiting for sidecar ready signal".to_string())
            .and_then(|r| r.map_err(|_| "sidecar closed before ready".to_string()))?;

        Ok(Self { stdin, child, dispatch, _reader })
    }

    // ── Reader loop (runs as a Tokio task) ──────────────────────────────────

    async fn reader_loop(
        mut reader: BufReader<tokio::process::ChildStdout>,
        dispatch: SharedDispatch,
        ready_tx: oneshot::Sender<()>,
    ) {
        let mut ready_sent = false;
        // Wrap in Option so we can consume (send) it exactly once.
        let mut ready_tx = Some(ready_tx);
        let mut line = String::new();

        loop {
            line.clear();
            match reader.read_line(&mut line).await {
                Ok(0) | Err(_) => break, // EOF or IO error
                Ok(_) => {}
            }

            let trimmed = line.trim();
            if trimmed.is_empty() {
                continue;
            }

            let v: Value = match serde_json::from_str(trimmed) {
                Ok(v) => v,
                Err(e) => {
                    eprintln!("[async_bridge] malformed JSON: {e}");
                    continue;
                }
            };

            // Ready signal
            if !ready_sent && v.get("ready") == Some(&Value::Bool(true)) {
                ready_sent = true;
                if let Some(tx) = ready_tx.take() {
                    let _ = tx.send(());
                }
                continue;
            }

            let id = match v.get("id").and_then(Value::as_str) {
                Some(s) => s.to_string(),
                None => continue,
            };

            // ── Chunk frame ─────────────────────────────────────────────────
            // A chunk frame carries "chunk" but neither "result" nor "error".
            if v.get("chunk").is_some()
                && v.get("result").is_none()
                && v.get("error").is_none()
            {
                let dp = dispatch.lock().await;
                if let Some(tx) = dp.streams.get(&id) {
                    let chunk = v.get("chunk").cloned().unwrap_or(Value::Null);
                    let _ = tx.send(Ok(chunk));
                }
                continue;
            }

            // ── Terminal / error frame ──────────────────────────────────────
            let is_error = v.get("error").is_some();
            let result = v.get("result").cloned();
            let error = v
                .get("error")
                .map(|e| parse_rpc_error(e));

            // Check stream dispatch first
            {
                let mut dp = dispatch.lock().await;
                if let Some(tx) = dp.streams.remove(&id) {
                    if let Some(err) = error {
                        let _ = tx.send(Err(err));
                    }
                    // Drop tx - this closes the channel signalling EOF.
                    drop(tx);
                    continue;
                }
            }

            // Regular call dispatch
            {
                let mut dp = dispatch.lock().await;
                if let Some(tx) = dp.calls.remove(&id) {
                    let payload = if is_error {
                        Err(error.unwrap_or(AsyncRpcError {
                            code: None,
                            message: "unknown error".to_string(),
                            traceback: None,
                            backtrace: None,
                        }))
                    } else {
                        Ok(result.unwrap_or(Value::Null))
                    };
                    let _ = tx.send(payload);
                } else {
                    eprintln!("[async_bridge] unknown response id: {id}");
                }
            }
        }

        // EOF reached - drain all pending waiters.
        if let Some(tx) = ready_tx.take() {
            let _ = tx.send(());
        }

        let mut dp = dispatch.lock().await;
        let exit_err = || AsyncRpcError {
            code: Some(-32000),
            message: "child process exited unexpectedly".to_string(),
            traceback: None,
            backtrace: None,
        };

        for (_, tx) in dp.calls.drain() {
            let _ = tx.send(Err(exit_err()));
        }
        for (_, tx) in dp.streams.drain() {
            let _ = tx.send(Err(exit_err()));
            // Drop tx to close the channel.
        }
    }

    // ── Public API ──────────────────────────────────────────────────────────

    /// Generate a unique call id.
    fn new_id() -> String {
        use std::sync::atomic::{AtomicU64, Ordering};
        static CTR: AtomicU64 = AtomicU64::new(1);
        format!("call-{}", CTR.fetch_add(1, Ordering::Relaxed))
    }

    /// Write a JSON line to the child's stdin.
    async fn write_line(&mut self, value: &Value) -> Result<(), String> {
        let mut line = value.to_string();
        line.push('\n');
        self.stdin
            .write_all(line.as_bytes())
            .await
            .map_err(|e| format!("stdin write error: {e}"))
    }

    /// Invoke `method` with `params` and wait for a single result value.
    pub async fn call(
        &mut self,
        method: &str,
        params: Value,
    ) -> Result<Value, AsyncRpcError> {
        let id = Self::new_id();

        let (tx, rx) = oneshot::channel::<Result<Value, AsyncRpcError>>();
        {
            let mut dp = self.dispatch.lock().await;
            dp.calls.insert(id.clone(), tx);
        }

        let req = serde_json::json!({
            "id": &id,
            "method": method,
            "params": params,
        });
        self.write_line(&req).await.map_err(|e| AsyncRpcError {
            code: None,
            message: e,
            traceback: None,
            backtrace: None,
        })?;

        rx.await.unwrap_or_else(|_| {
            Err(AsyncRpcError {
                code: Some(-32000),
                message: "reader task dropped before responding".to_string(),
                traceback: None,
                backtrace: None,
            })
        })
    }

    /// Invoke `method` with `params` and return an [`AsyncStream`] that yields
    /// each chunk as it arrives.  The stream ends when the sidecar sends a
    /// terminal frame.
    pub async fn stream(
        &mut self,
        method: &str,
        params: Value,
    ) -> Result<AsyncStream, AsyncRpcError> {
        let id = Self::new_id();

        let (tx, rx) = mpsc::unbounded_channel::<Result<Value, AsyncRpcError>>();
        {
            let mut dp = self.dispatch.lock().await;
            dp.streams.insert(id.clone(), tx);
        }

        let req = serde_json::json!({
            "id": &id,
            "method": method,
            "params": params,
            "stream": true,
        });
        self.write_line(&req).await.map_err(|e| AsyncRpcError {
            code: None,
            message: e,
            traceback: None,
            backtrace: None,
        })?;

        Ok(AsyncStream { rx })
    }

    /// Convenience: send `__ping__` and return the pong response.
    pub async fn ping(&mut self) -> Result<Value, AsyncRpcError> {
        self.call("__ping__", serde_json::json!({})).await
    }

    /// Gracefully shut down the bridge: close stdin and wait for the child.
    pub async fn close(mut self) {
        // Drop stdin so the child sees EOF and can shut down cleanly.
        drop(self.stdin);
        let _ = self.child.wait().await;
    }

    /// Kill the child process immediately and wait for it to exit.
    pub async fn kill(mut self) {
        let _ = self.child.kill().await;
        let _ = self.child.wait().await;
    }
}
